de-RFC: Remove unsized_locals (#3829)
rust-lang/rfcs: text/3829-de-rfc-unsized-locals.md

- Feature Name: `unsized_locals`
- Start Date: 2025-06-02
- RFC PR: [rust-lang/rfcs#3829](https://github.com/rust-lang/rfcs/pull/3829)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

_Following the great success of the [type ascription de-RFC], we now present the next one in the series!_

# Summary
[summary]: #summary

> Unsized locals ([RFC 1909], called "unsized rvalues" originally)

Author: 0e0b1a55f6825d4cad326c4a743fcbe1cf741e51

has been a merged RFC for eight years with no clear path to stabilization.

> There is still a very large gap in the implementation in rustc that hasn't been adressed for many years and there are several language problems with this feature. 

Reviewer: ```suggestion
There is still a very large gap in the implementation in rustc that hasn't been addressed for many years and there are several language problems with this feature. 
```

This RFC intends to advocate for the feature being removed entirely with a fresh RFC being necessary to add a similiar feature again.

# Demotivation
[demotivation]: #demotivation

The `unsized_locals` feature is simple to explain on the surface: local variables no longer have to be `Sized`.

```rust
#![feature(unsized_locals)]
fn main() {
    let x = *Box::from("hello, world!");
}
```

This will dynamically allocate space on the stack for the string.

> C has a similiar feature, [`alloca`] and variable length arrays (VLA), the latter having been made optional to implement in C11.

Reviewer: ```suggestion
C has a similar feature, [`alloca`] and variable length arrays (VLA), the latter having been made optional to implement in C11.
```

## Lack of Proper Implementation

This feature has never been properly implemented in rustc.
The variable length array form proposed in the RFC still doesn't exist at all, only the unsized local variable does.
It is implemented in the type checker and codegen, but lacking MIR semantics and therefore unimplemented in the compile time function evaluator.
This is very significant, as MIR semantics govern how the feature should behave precisely in the first place.

> Without them, they cannot work in `const` and optimizations are likely broken around them.

Reviewer 2: I would like to understand the “MIR semantics” issues in more detail. What exactly would the compiler team need to see before it would become possible to move forward? https://github.com/rust-lang/compiler-team/issues/630 says:

> The [MIR semantics] that const eval used to implement - before support was removed - were known to be awkward and surprising.

But doesn’t link to any of that discussion. It’s hard for people to step up and address the problems, without a clear summary of what they are.

Reviewer: Even then, a detailed summary of the problems would also have to be updated for changes in MIR semantics since then, no? Which may be potentially quite a bit.

Reviewer 2: I found https://github.com/rust-lang/rust/issues/48055#issuecomment-1837424794 which seems to be what I was looking for.

Author: I added a link to the comment in 5bc11846302f5fe76326f065c33b554339979e91

Reviewer 2: I started a discussion about fixing the MIR on Zulip: https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/MIR.20semantics.20for.20unsized.20locals

Because of this lack of implementation quality, the `unsized_locals` feature was already accepted for removal from rustc two years ago in [MCP 630].
This removal hasn't yet been implemented.

> ## Implicit Danger

Reviewer 2: Consider this bog-standard generic function:

```rust
fn foo(arg: impl SomeTrait) {
     // do stuff with `arg`, including store it in locals
}
```

To my mind, if the above is considered totally mundane and not-scary, then this:

```rust
fn foo_with_dynamic_dispacth(arg: dyn SomeTrait) {
     // do stuff with `arg`, including store it in locals
}
```

Should not be scary either. The risk of `dyn SomeTrait` being huge and blowing the stack, is no different from the risk of `impl SomeTrait` being huge and blowing the stack.

> ## Implicit Danger

Reviewer: The discussion of `dyn Trait` is a separate matter. I'm resolving this suggestion.

While lack of implementation quality is a sign of lack of interest for the feature, it is not the primary reason for this RFC,

> which is purely about the langauge design of the feature.

Reviewer 3: ```suggestion
which is purely about the language design of the feature.
```

The original RFC was very short, and especially short on motivation and rationale for the design.

> Dynamic stack allocation in geneneral has a rather significant downside: it makes it easy to accidentally overflow the stack if you allocate a lot.

Reviewer: ```suggestion
Dynamic stack allocation in general has a rather significant downside: it makes it easy to accidentally overflow the stack if you allocate a lot.
```

Stacks are usually rather small (on the order of few megabytes, depending on the platform), which means that dynamically allocating user-controlled input on the stack is often rather dangerous.

> While stack overflows are not considered memory unsafe by Rust, they still cause crashes which can lead to denial of service vulnerabilities and unrealiableness in general.

Reviewer: ```suggestion
While stack overflows are not considered memory unsafe by Rust, they still cause crashes which can lead to denial of service vulnerabilities and unreliability in general.
```

Dynamic stack allocation as currently implemented interacts especially poorly with loops.
Allocations are not freed until the function returns, so the following example overflows the stack:

```rust
#![feature(unsized_locals)]
fn main() {
    let s = "A".repeat(1000);
    for i in 0..1000000 {
        let x: str = *Box::from(s.as_str());
        std::hint::black_box(&s);
    }
}
```

> There are ways around this (rewinding the stack pointer at the end of the loop to free up the memory), but they are not currently implemented.

Reviewer 2: [RFC 1909](https://rust-lang.github.io/rfcs/1909-unsized-rvalues.html) does not specify “unsized locals in a loop leak a bunch of memory for no reason”. It’s not part of the “design,” it’s a codegen bug. We know how to fix it, that fix would need to happen before stabilization.

> There are ways around this (rewinding the stack pointer at the end of the loop to free up the memory), but they are not currently implemented.

Reviewer: ...I think, to actually implement the suggestion, we would have to allow every `let`-binding of an alloca that is nested inside a loop to turn that loop into a closure. The closure would have to capture all the values referenced inside the loop and then return a value representing the control flow state to decide whether we must e.g. `return`. And then we would have to hope that the dynamic allocation then doesn't e.g. get inlined?There's some possible alternatives that could use exception handling as an implementation if it performed better, but would otherwise be similar in concept.

While I am obviously hypothesizing, based on insufficient information, I do not expect any such implementation to not be a significant-enough optimization barrier that itself would detract from the intended usage. And if it isn't, it seems it would risk the optimization actually reintroducing the stack overflow risk.

> There are ways around this (rewinding the stack pointer at the end of the loop to free up the memory), but they are not currently implemented.

Reviewer: huh, I missed those instructions while specifically looking for instructions like that. hurf.

> There are ways around this (rewinding the stack pointer at the end of the loop to free up the memory), but they are not currently implemented.

Reviewer 2: We’ve all been there…

Dynamic stack allocation also has its upsides.
It is generally faster than heap allocation and can therefore improve performance in cases where the previously mentioned downsides are not a concern.

> Therefore, this RFC is not necessarily a rejection of the idea of dynamic stack allocation, but merely the way the `unsized_locals` feature exposes it.

Reviewer 2: I agree that we still need to figure out the best way to expose the feature, but I think that the best way to do that is to implement it properly on nightly and enable experimentation. Right now, we have vague concerns about footguns and blowing the stack, but without concrete user feedback, it’s very hard to know how severe the issue is or how the problem could be mitigated.

Reviewer 2: E.g. maybe lints could address the danger. Or maybe all unsized locals should be annotated in some way (e.g. `let #[unsized] foo = …`). We can’t try those things out unless we have a proper implementation.

Reviewer: Then go implement it if you think it's such a good idea.

Author: No significant experimentation has happened for the past **eight** years. I wouldn't hold my beer.

Reviewer: Right, it's actually the middle-end, I think?

Reviewer 2: > No significant experimentation has happened for the past **eight** years.

Yes, because the middle/back-end implementation is broken, scaring nightly users away. If the compiler team would rather strip out that broken implementation than continue to let it rot, I have no objection. But https://github.com/rust-lang/compiler-team/issues/630 is enough for that. IIUC this de-RFC would require anyone motivated to implement the feature properly, to go through the full RFC process first. I think that’s going too far. My preference would be to “downgrade” #1909 to an eRFC, instead of unaccepting it completely.

Reviewer: Changing it to an "experimental" RFC will not change that right now it is a gravestone of a feature that no one wants to proceed on. Yes, probably someone *should* go through the RFC process again, because right now there's not even a draft of a proposal for how it may be implemented in rustc's semantic layers.

Reviewer 2: >  it is a gravestone of a feature that no one wants to proceed on

I wouldn’t say that. For example, someone recently put in the effort to fix the longstanding issue with unsized locals being misaligned. Clearly there are *some* people still interested in this.

Reviewer: I believe that what @tmiasko was touching on there involved fixes to parts of the code that were shared between `unsized_locals` and `unsized_fn_params`, because code that accepts `unsized_fn_params` sometimes, due to the way codegen works, creates unsized temporaries, i.e. locals. That then requires handling that for `unsized_fn_params`. Our current rude hack that addresses this involves blocking inlining for `unsized_fn_params`, since that's one of the conditions to reach that.

Also see:
- https://github.com/rust-lang/rust/issues/111355
- https://github.com/rust-lang/rust/issues/111175

Which is where we get to the major argument of this RFC: The `unsized_locals` feature integrates far too naturally into Rust.
This makes the feature very **easy to use**, and especially **easy to use accidentally**.

As previously mentioned, dynamic stack allocation can be dangerous and should not be used lightly.
It's an advanced optimization feature that is best left untouched by default.
As such, it behaves similarly to `unsafe` (but is not actually `unsafe`).
With `unsized_locals`, the use of dynamic stack allocation is completely implicit.

> When you create an unsized local, it is often not obvious that dynamic stack allocation is happening.

Reviewer: In practice, I rarely see recursive code in Rust, but I do see `let` bindings a lot, so this doesn't seem like a very strong argument.

> When you create an unsized local, it is often not obvious that dynamic stack allocation is happening.

Reviewer 2: You do see function calls a lot, though. And function calls can be accidentally recursive, just as locals can be accidentally unsized

> When you create an unsized local, it is often not obvious that dynamic stack allocation is happening.

Reviewer 4: On one hand the stack usage by recursion is O(n) --- you need to call itself too many times before blowing up the stack,
and alloca is O(1) --- you just need to one wrong `let` to crash the entire program, so `unsized_locals` is indeed much more dangerous than recursion.

On the other hand, the same applies for *sized* locals
```rust
#[expect(clippy::large_stack_arrays)] // <- this is pedantic level lint, not enabled by default
let a = [2; 1_000_000_000_000];
assert_eq!(a.last(), Some(&2));
```
(that 1_000_000_000_000 can be hidden from the reader through const calculations like `size_of::<T>() * 100` or some multi-dimensional arrays)

For sure this is not *dynamic* stack allocation, with alloca the compiler cannot verify that the length is bounded by some number and thus cannot deterministically issue that equivalent of `clippy::large_stack_arrays`, but if `[2; 1_000_000_000_000]` is allowed to be safe there seems no reason to say alloca behaves like unsafe. :thinking:

> When you create an unsized local, it is often not obvious that dynamic stack allocation is happening.

Reviewer: this is true, but `[2; 1_000_000_000_000]` seems slightly more obvious in being a gunshot at your foot.

@Jules-Bertholet I'm not going to contest that you can have accidental recursion, but we're still talking about several orders of magnitude less in frequency than "`let` binding in a loop". Indeed, the fact that loops (or iterator combinators) are so often the more ergonomic choice in Rust is part of why I see recursion less.

> When you create an unsized local, it is often not obvious that dynamic stack allocation is happening.

Reviewer 4: @Jules-Bertholet `let local = *this_slice_might_be_huge` won't compile because `[u32]` is not Copy.

This works though:

```rust
#![feature(unsized_locals)]

fn g(a: Box<[u32]>) -> u32 {
    let b = *a;
    // thread '<unknown>' has overflowed its stack
    // fatal runtime error: stack overflow, aborting
    *b.last().unwrap()
}

fn main() {
    let a = vec![2; 10_000_000].into_boxed_slice();
    assert_eq!(g(a), 2);
}
```

Compare with `unsized_fn_params` which runs fine

```rust
#![feature(unsized_fn_params)]

fn g(a: [u32]) -> u32 {
    *a.last().unwrap()
}

fn main() {
    let a = vec![2; 10_000_000].into_boxed_slice();
    assert_eq!(g(*a), 2);
}
```

In the example from the start, we need to be aware of all the involved types (which are often inferred in practice) to know that this is a potentially problematic unsized local that we have to audit more carefully instead of a normal sized local.

> Especially around strings, which are often user-controlled, this easily lead to accidental dangerous situations.

Reviewer: ```suggestion
Especially around strings, which are often user-controlled, this easily leads to accidentally-dangerous situations.
```

> Rusts strings, reference types, and `Sized` are a part of the language that can often be hard to understand for beginners coming from garbage collected languages.

Reviewer: ```suggestion
Rust's strings, reference types, and `Sized` are a part of the language that can often be hard to understand for beginners coming from garbage-collected languages.
```

With this RFC, the fate of `unsized_locals` is sealed and it becomes clear to anyone what the state of the feature is.

The intent of this RFC is not to proposed an alternative way to solve dynamic stack allocation,
but there are some listed alternatives here that may be considered in the future if desired.
While this RFC doesn't explicitly encourage people to revisit this topic, it may result in new activity around dynamic stack allocation in Rust.

[RFC 1808] proposed an `alloca` function, which was rejected because `alloca` does not really behave like a function.

[RFC 1808] then changed to propose the the VLA syntax instead.
It was rejected in favor of more general unsized values, which culminated in [RFC 1909].

The [`alloca` crate](https://crates.io/crates/alloca) implements dynamic stack allocation via a closure indirection and FFI with C.

The `unsized_fn_params` feature doesn't suffer from the same problems as `unsized_locals` and will still be kept around.
It is independent of this RFC.

# Posterior art
[posterior-art]: #posterior-

The best prior art for this removal is of course the inspiration for the de-RFC format, the [type-ascription de-RFC].

> [MCP 630] can also be seen as prior art

Reviewer: ```suggestion
[MCP 630] can also be seen as prior art.
```

# Unresolved answers
[unresolved-answers]: #unresolved-answers

None

# Future probabilities
[future-probabilities]: #future-probabilities

In the future, dynamic stack allocation may be re-added to Rust via some other feature that solves the explicitness problems outlined in the motivation.

Alternatively, it could be decided, explicitly or implicitly through inaction, that dynamic stack allocation is not a fit for Rust and will not be added.

> [type-ascription de-RFC]: https://rust-lang.github.io/rfcs/3307-de-rfc-type-ascription.html

Reviewer 4: ```suggestion
[type ascription de-RFC]: https://rust-lang.github.io/rfcs/3307-de-rfc-type-ascription.html
```

- Feature Name: `unsized_locals`
- Start Date: 2025-06-02
- RFC PR: [rust-lang/rfcs#3829](https://github.com/rust-lang/rfcs/pull/3829)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

_Following the great success of the [type ascription de-RFC], we now present the next one in the series!_

# Summary
[summary]: #summary

Unsized locals ([RFC 1909], called "unsized rvalues" originally)
has been a merged RFC for eight years with no clear path to stabilization.

There is still a very large gap in the implementation in rustc that hasn't been addressed for many years and there are several language problems with this feature. 

> This RFC intends to advocate for the feature being removed entirely with a fresh RFC being necessary to add a similiar feature again.

Reviewer 5: ```suggestion
This RFC intends to advocate for the feature being removed entirely with a fresh RFC being necessary to add a similar feature again.

Note that the acceptance of this RFC should not be taken as judgement on a future RFC. A fresh RFC with a new design would be required, and this RFC may provide input on design constraints for such an RFC, but this should not be taken as pre-rejecting such an RFC.
```

As a _dangerous_ feature, dynamic stack allocation must be explicit and obvious - and `unsized_locals` makes it implicit _by design_.
Therefore, `unsized_locals` must go.

`unsized_locals` is not the only feature that can cause unbounded stack allocation and eventually lead to stack overflow in Rust, one has to look no further than Dijkstra's favorite: recursion.
A program with unbounded recursion is easily found and fixed, but especially recursive parsers suffer from similar problems where user input can cause stack overflows.
What makes recursion different from dynamic stack allocation?
The main difference here is that recursion is a lot harder to do on accident.
When recursion is used, it is usually used on purpose, and while sometimes the potential for stack overflows is overlooked, the general feature is usually used on purpose.
That said, recursion can certainly be dangerous in some contexts, but prior existing features are not a good reason to introduce more ways to blow the stack.

> The Linux kernel has spent a lot of time on getting VLAs (C's cousin to `unsized_locals`) [removed from the codebase](https://www.phoronix.com/news/Linux-Kills-The-VLA).

Reviewer 2: > Trivially `dyn AsRef<[u8]>` is not materially better than `[u8]` on the stack

And `impl<AsRef<[u8]>>` is also going to be equally bad, but we allow that today, and it doesn’t cause problems.

# Drawforwards
[drawforwards]: #drawforwards

This feature has the previously mentioned performance upsides that users could profit from if it was stabilized.
But this benefit applies to other ways to expose dynamic stack allocation too, and other ways are likely to be easier to implement correctly and stabilize.

# Irrationale and alternatives
[irrationale-and-alternatives]: #irrationale-and-alternatives

If nothing is done on the language side, the feature will likely still be removed from the compiler.
This puts the feature into a really bad position, but it may be readded in the future if someone desires.
With this RFC, the fate of `unsized_locals` is sealed and it becomes clear to anyone what the state of the feature is.

The intent of this RFC is not to proposed an alternative way to solve dynamic stack allocation,
but there are some listed alternatives here that may be considered in the future if desired.
While this RFC doesn't explicitly encourage people to revisit this topic, it may result in new activity around dynamic stack allocation in Rust.

[RFC 1808] proposed an `alloca` function, which was rejected because `alloca` does not really behave like a function.

> [RFC 1808] then changed to propose the the VLA syntax instead.

Reviewer 6: ```suggestion
[RFC 1808] then changed to propose the VLA syntax instead.
```

It was rejected in favor of more general unsized values, which culminated in [RFC 1909].

The [`alloca` crate](https://crates.io/crates/alloca) implements dynamic stack allocation via a closure indirection and FFI with C.

The `unsized_fn_params` feature doesn't suffer from the same problems as `unsized_locals` and will still be kept around.
It is independent of this RFC.

# Posterior art
[posterior-art]: #posterior-

The best prior art for this removal is of course the inspiration for the de-RFC format, the [type ascription de-RFC].

> [MCP 630] can also be seen as prior art.

Reviewer 6: Should we add a reference to this de-RFC to the type ascription de-RFC as posterior art?
